README.rst
sublime-harness
===============

.. image:: https://travis-ci.org/twolfson/sublime-harness.png?branch=master
   :target: https://travis-ci.org/twolfson/sublime-harness
   :alt: Build Status

Run Python in Sublime Text from outside of Sublime Text

``sublime-harness`` was built to allow for execution of arbitrary Python within the context of `Sublime Text`_. It is also part of the `Sublime plugin tests`_ framework.

.. _`Sublime Text`: http://sublimetext.com/
.. _`Sublime plugin tests`: https://github.com/twolfson/sublime-plugin-tests

    Currently, only Linux is supported but OSX and Windows support are planned.

Deprecation notice
==================
We have decided to deprecate ``sublime-harness`` in favor of `randy3k/UnitTesting`_.

It has greater platform support and a less brittle design for local development.

https://github.com/randy3k/UnitTesting

.. _`randy3k/UnitTesting`: https://github.com/randy3k/UnitTesting

Getting Started
---------------
Install the module with: ``pip install sublime_harness``

.. code:: python

    import os, time
    from sublime_harness import Harness

    harness = Harness()
    script = """
    import sublime

    # Harness will run the `run` function
    def run():
        with open('/tmp/hi', 'w') as f:
            f.write('Hello World!')
        sublime.run_command('exit')"""
    harness.run(script)

    # Wait for our file to exist (Sublime Text is forked and not synchronous)
    output_file = '/tmp/hi'
    while (not os.path.exists(output_file) or
           os.stat(output_file).st_size == 0):
        time.sleep(0.1)

    # Read our data
    with open(output_file) as f:
      print f.read()  # Hello World!

Documentation
-------------
``sublime_harness`` provides the ``Harness`` class for all your bootstrapping needs.

`Sublime Text`_ will be resolved via `sublime-info`_, which allows for overriding via environment variables.

.. _`sublime-info`: https://github.com/twolfson/sublime-info

Harness.__init__
^^^^^^^^^^^^^^^^
.. code:: python

    Harness()
    """Generate a new Harness for Sublime Text

    When initialized, `Harness` allocates a directory (currently,
    same for all harnesses) for your script.
    """

Harness.directory
^^^^^^^^^^^^^^^^^
.. code:: python

    harness.directory
    """Directory where `run` will be execute

    If you would like to load relative modules, they should be copied to this directory."""

Harness.run
^^^^^^^^^^^
.. code:: python

    harness.run(script)
    """Python to execute within the context of Sublime Text

    **YOU MUST CLEAN UP AFTER RUNNING THIS METHOD VIA `close`**

    You can only run one harness at a time due to the lack of namespacing.

    :param script: Python to execute within Sublime Text
    :type script: str
    """

Harness.close
^^^^^^^^^^^^^
.. code:: python

    harness.close()
    """Cleans up harness files"""

Examples
--------
As mentioned within ``Harness.dictionary``, external files can be loaded relatively to the script. This is an example of how to set up and use them.

.. code:: python

    # Set up a new harness
    import os, time
    from sublime_harness import Harness
    harness = Harness()

    # Copy over a local file to the directory
    dest_hello_path = harness.directory + '/hello.py'
    with open(dest_hello_path, 'w') as f:
      f.write('hello = "World!"')

    # Generate and run our temporary task
    script = """
    import os
    import sublime

    try :
        from hello import hello  # ST 2 compatible
    except:
        from .hello import hello  # ST 3 compatible

    def run():
        with open('/tmp/hi-directory', 'w') as f:
            f.write(hello)
        sublime.run_command('exit')
    """

    harness.run(script)
    output_file = '/tmp/hi-directory'
    while (not os.path.exists(output_file) or
           os.stat(output_file).st_size == 0):
        time.sleep(0.1)

    # Grab the file output
    with open(output_file) as f:
        print f.read() # 'World!'

    # Remove the plugin and our file
    harness.close()
    os.unlink(dest_hello_path)

Contributing
------------
In lieu of a formal styleguide, take care to maintain the existing coding style. Add unit tests for any new or changed functionality. Test via ``nosetests``.

Donating
--------
Support this project and `others by twolfson`_ via `gittip`_.

.. image:: https://rawgithub.com/twolfson/gittip-badge/master/dist/gittip.png
   :target: `gittip`_
   :alt: Support via Gittip

.. _`others by twolfson`:
.. _gittip: https://www.gittip.com/twolfson/

Unlicense
---------
As of Oct 16 2013, Todd Wolfson has released this repository and its contents to the public domain.

It has been released under the `UNLICENSE`_.

.. _UNLICENSE: https://github.com/twolfson/sublime-harness/blob/master/UNLICENSE


CHANGELOG.rst
sublime-harness changelog
=========================
1.0.0 - Deprecated ``sublime-harness``

0.4.2 - Added foundry for release

0.4.1 - Updated all ``rawgithub.com`` references to ``raw.github.com``

0.4.0 - Added ``Harness`` to ``__init__.py`` to allow for ``from sublime_harness import Harness``

0.3.0 - Fixed up environment variable and tests for Vagrant

0.2.0 - Updated run order of tests for faster feedback loops against Travis CI

0.1.1 - Corrected invalid headings in README

0.1.0 - Initial release


test.sh
SUBLIME_TEXT_PATH=/opt/sublime_text/sublime_text nosetests --nocapture --verbose --stop

test/http_logger.py
# TODO: Consider using https://pypi.python.org/pypi/wsgilog/0.3
import BaseHTTPServer


# Create a server which logs POST body to stdout
class LogServer(BaseHTTPServer.BaseHTTPRequestHandler):
    def do_POST(self):
        len = int(self.headers.getheader('content-length'))
        print self.rfile.read(len)
        self.send_response(200, 'Logged to stdout')


# Create our server
app = BaseHTTPServer.HTTPServer(('', 7070), LogServer)
app.serve_forever()


test/sublime-harness_test.py
import os
import shutil
import tempfile
import time
import unittest
import subprocess

import sublime_info

import sublime_harness

# Set up constants
__dir__ = os.path.dirname(os.path.abspath(__file__))

# Outline tests
"""
sublime-harness
    running arbitrary Python  # write to disk
        executes the code
    running Sublime Text specific code  # open a view, write content, get content, write fetched content to disk
        executes within Sublime Text

# Intermediate test
sublime-harness
    runs in the `directory`  # write files to `directory`, assert they can be loaded from

# Edge case
sublime-harness
    will raise if code does not contain a `run` method
Sublime Text
    running after running a sublime-harness
        does not re-execute harness code

# TODO: If there is an error launching Sublime Text, proliferate it
# Just not stderr (fucking headless + xvfb)
# which means status code
"""


class TestSublimeHarness(unittest.TestCase):
    def setUp(self):
        self.output_file = tempfile.mkstemp()[1]
        self.harness = sublime_harness.Harness()

    def tearDown(self):
        os.unlink(self.output_file)

        # If we are autokilling, wait for all Sublime's to close
        if os.environ.get('SUBLIME_AUTO_KILL'):
            while self._sublime_is_running():
                time.sleep(0.1)

    def _wait_for_output_file(self):
        output_file = self.output_file
        while (not os.path.exists(output_file) or
               os.stat(output_file).st_size == 0):
            time.sleep(0.1)

    def _sublime_is_running(self):
        child = subprocess.Popen(['ps', 'ax'], stdout=subprocess.PIPE)
        ps_list = str(child.stdout.read())
        child.kill()
        sublime_is_running = False
        sublime_path = sublime_info.get_sublime_path()
        sublime_command = os.path.basename(sublime_path)
        for process in ps_list.split('\n'):
            if sublime_command in process:
                sublime_is_running = True
                break
        return sublime_is_running

    def test_running_arbitrary_python(self):
        # TODO: This test is useless due to sublime self-kill
        # Generate and run our temporary task
        plugin_str = open(__dir__ + '/test_files/arbitrary.py').read() % self.output_file
        self.harness.run(plugin_str)
        self._wait_for_output_file()

        # Grab the file output
        with open(self.output_file) as f:
            self.assertIn('hello world', f.read())

        # Remove the plugin
        self.harness.close()

    def test_running_st_python(self):
        # Generate and run our temporary task
        plugin_str = open(__dir__ + '/test_files/st_python.py').read() % self.output_file
        self.harness.run(plugin_str)
        self._wait_for_output_file()

        # Grab the file output
        with open(self.output_file) as f:
            self.assertIn('Packages', f.read())

        # Remove the plugin
        self.harness.close()

    def test_run_in_directory(self):
        # Copy over a local file to the directory
        dest_hello_path = self.harness.directory + '/hello.py'
        shutil.copy(__dir__ + '/test_files/hello.py', dest_hello_path)

        # Generate and run our temporary task
        plugin_str = open(__dir__ + '/test_files/directory.py').read() % self.output_file
        self.harness.run(plugin_str)
        self._wait_for_output_file()

        # Grab the file output
        with open(self.output_file) as f:
            self.assertEqual('world', f.read())

        # Remove the plugin and our file
        self.harness.close()
        os.unlink(dest_hello_path)

    def test_assert_run(self):
        assert_run_plugin = open(__dir__ + '/test_files/missing_run.py').read()
        self.assertRaises(Exception, self.harness.run, assert_run_plugin)

    # DEV: Namez with zz to always run last (to see if lower level issues or this test are the problem)
    def test_zz_prevent_multiple_runs(self):
        # DEV: In Vagrant + ST3, this takes 15 seconds to run
        # Generate and run our temporary task
        plugin_str = open(__dir__ + '/test_files/timestamp.py').read() % self.output_file
        self.harness.run(plugin_str)
        self._wait_for_output_file()

        # Grab the file output
        f = open(self.output_file)
        orignal_timestamp = f.read()
        f.close()
        self.assertNotEqual(orignal_timestamp, '')

        # Remove the plugin
        self.harness.close()

        # Launch sublime again
        if os.environ.get('TRAVIS') or os.environ.get('VAGRANT'):
            # DEV: Currently `subl` in Travis launches a fork from a bash shell meaning it uses a different PID which is destroyed.
            # TODO: We should add `/opt/sublime_text_2/sublime_text` as a primary check point for subl
            if sublime_info.get_sublime_version() < 3000:
                child = subprocess.Popen(['/opt/sublime_text_2/sublime_text', '--class=sublime-text-2'], stderr=subprocess.PIPE)
            else:
                child = subprocess.Popen(['/opt/sublime_text/sublime_text', '--wait'])
        else:
            child = subprocess.Popen([sublime_info.get_sublime_path()], stderr=subprocess.PIPE)

        # Wait for a bit
        time.sleep(1)

        # Kill Sublime
        child.kill()

        # Assert file has not changed
        f = open(self.output_file)
        new_timestamp = f.read()
        f.close()
        self.assertEqual(orignal_timestamp, new_timestamp)


sublime_harness/logger.py
import os
try:
    from urllib.request import urlopen
    from urllib.error import URLError
except ImportError:
    from urllib2 import urlopen, URLError

# TODO: Consider using a proper logger
class IgnorantLogger(object):
    def _log(self, level, msg):
        pass

    def debug(self, msg):
        self._log('debug', msg)

    def info(self, msg):
        self._log('info', msg)

    def warn(self, msg):
        self._log('warn', msg)


# TODO: Robustify this with port and address
class HttpLogger(IgnorantLogger):
    def _log(self, level, msg):
        try:
            # TODO: This is horrible string interpolation
            req = urlopen('http://localhost:7070', data='level={}&msg={}'.format(level, msg))
            req.read() # DEV: Stop open pipes from occurring
        except URLError:
            pass


# TODO: This is really bad technique. We should be doing this based off of a factory.
# TODO: Then, inside of the place where we load the program we can use env var
logger = HttpLogger() if os.environ.get('HARNESS_LOGGER') else IgnorantLogger()


sublime_harness/sublime_harness.py
# Load in core dependencies
import ast
import os
import shutil
import subprocess

# Load in 3rd party dependencies
import sublime_info

# Load local dependencies
from .logger import logger

# Set up constants
__dir__ = os.path.dirname(os.path.abspath(__file__))

# TODO: init launcher needs that file safeguard to deal with people who don't clean up
# When init plugin is loading, check for `.abc`.
# If it exists, continue executing. Once launched, remove it.
# During the timeout check, keep on verifying it exists (we don't want double load race conditions)
# Otherwise, do nothing.


class Harness(object):
    sublime_path = sublime_info.get_sublime_path()
    sublime_command = os.path.basename(sublime_path)

    def ensure_directory(self):
        # If the harness directory does not exist, create it
        if not os.path.exists(self.directory):
            os.makedirs(self.directory)

    def install_command_launcher(self):
        # If command launcher doesn't exist, copy it
        # TODO: Verify this works in Windows
        orig_command_path = os.path.join(__dir__, 'launchers/command.py')
        dest_command_path = os.path.join(self.directory, 'command_launcher.py')
        if not os.path.exists(dest_command_path):
            shutil.copyfile(orig_command_path, dest_command_path)
        else:
        # Otherwise...
            # If there are updates for command launcher
            expected_command = None
            with open(orig_command_path) as f:
                expected_command = f.read()
            actual_command = None
            with open(dest_command_path) as f:
                actual_command = f.read()
            if expected_command != actual_command:
                # Update the file
                shutil.copyfile(orig_command_path, dest_command_path)

                # and notify the user we must restart Sublime
                raise Exception('We had to update the sublime-harness plugin. You must close or restart Sublime to continue testing.')


    def install_init_launcher(self):
        # Clean up any past instances of init launcher
        self.remove_init_launcher()

        # Install a new one
        # TODO: Verify this doesn't have any double invocation consequences
        orig_command_path = os.path.join(__dir__, 'launchers/init.py')
        shutil.copyfile(orig_command_path, self.init_launcher_path)

    def remove_init_launcher(self):
        # If the init launcher exists, delete it
        if os.path.exists(self.init_launcher_path):
            os.unlink(self.init_launcher_path)

    def __init__(self):
        # TODO: Require namespace for harness directory
        # Guarantee the harness directory exists
        self.directory = os.path.join(sublime_info.get_package_directory(), 'sublime-harness-tmp')
        self.ensure_directory()

        self.init_launcher_path = os.path.join(self.directory, 'init_launcher.py')

        # Save defaults
        self.close_called = True

    def run(self, script):
        # TODO: Add safeguard for only running once at a time

        # Assert we are provided a run function
        has_run_fn = False
        for node in ast.iter_child_nodes(ast.parse(script)):
            try:
                if (node.__class__.__name__ == 'FunctionDef' and node.name == 'run'):
                    has_run_fn = True
                    break
            except:
                pass

        # If there was no run function, complain and leave
        if not has_run_fn:
            raise Exception('"run" function was not found in "script": %s' % script)

        # Output test to directory
        f = open(os.path.join(self.directory, 'plugin.py'), 'w')
        f.write(script)
        f.close()

        # TODO: These commands should go in a launching harness
        # If we are running Sublime Text 3 and it has not yet started, use `init`
        running_via_init = False
        if sublime_info.get_sublime_version() >= 3000:
            # TODO: Use tasklist for Windows
            # Get process list
            child = subprocess.Popen(['ps', 'ax'], stdout=subprocess.PIPE)
            ps_list = str(child.stdout.read())

            # Kill the child
            child.kill()

            # Determine if Sublime Text is running
            # TODO: This could be subl, sublime_text, or other
            sublime_is_running = False
            for process in ps_list.split('\n'):
                if self.sublime_command in process:
                    sublime_is_running = True
                    break

            # If sublime isn't running, use our init trigger
            logger.debug('Current process list: %s' % ps_list)
            if not sublime_is_running:
                # Install the init trigger
                self.install_init_launcher()

                # and launch sublime_text
                logger.info('Launching %s via init' % self.sublime_path)
                subprocess.call([self.sublime_path])

                # Mark the init to prevent double launch
                running_via_init = True

        # Save running_via_init info
        self.running_via_init = running_via_init

        # If we are not running via init
        if not running_via_init:
            # Guarantee the command launcher exists
            self.install_command_launcher()

            # Invoke the launcher command
            # TODO: Use real namespace
            logger.info('Launching %s via --command' % self.sublime_path)
            subprocess.call([self.sublime_path, '--command', 'sublime_harness_command_launcher_namespace'])

        # Mark close as not called
        self.close_called = False

    def close(self):
        # TODO: When we get to generic directories, clean them up
        # If we were running via init, clean it up
        if self.running_via_init:
            self.remove_init_launcher()


sublime_harness/launchers/command.py
# Load in core dependencies
import code
import os
import sublime
import sublime_plugin

# Set up constants
__dir__ = os.path.dirname(os.path.abspath(__file__))


class SublimeHarnessCommandLauncherNamespaceCommand(sublime_plugin.ApplicationCommand):
    def run(self):
        # On every run, re-import the test class
        # DEV: Sublime Text does not recognize changes to command.py.
        # DEV: Once it is loaded and run once via CLI, it is locked in memory until Sublime Text is restarted
        filepath = __dir__ + '/plugin.py'
        namespace = 'sublime-harness-tmp'
        is_sublime_2k = sublime.version() < '3000'
        plugin_dict = {
            '__dir__': __dir__,
            '__file__': filepath,
            '__name__': 'plugin' if is_sublime_2k else '%s.plugin' % namespace,
            '__package__': None if is_sublime_2k else namespace,
            '__builtins__': __builtins__,
        }

        # DEV: In Python 2.x, use execfile. In 3.x, use compile + exec.
        # if getattr(__builtins__, 'execfile', None):
        if is_sublime_2k:
            execfile(filepath, plugin_dict, plugin_dict)
        else:
            f = open(filepath)
            script = f.read()
            interpretter = code.InteractiveInterpreter(plugin_dict)
            interpretter.runcode(compile(script, filepath, 'exec'))
        plugin_dict['run']()


sublime_harness/launchers/init.py
# Load in core dependencies
import code
import os
import sys
import sublime
import sublime_plugin
import traceback

try:
    from urllib.request import urlopen
except ImportError:
    from urllib2 import urlopen

urlopen('http://localhost:7070', data=b'init is running')


# Set up constants
__dir__ = os.path.dirname(os.path.abspath(__file__))
run_placeholder_path = os.path.join(__dir__, '.init_launcher_can_run')


# Set up hook for set timeout loop
plugin_host_loaded = False


class SublimeHarnessInitLauncherNamespaceCommand(sublime_plugin.ApplicationCommand):
    def run(self):
        try:
            self._run()
        except Exception:
            exc_type, exc_value, exc_traceback = sys.exc_info()
            err = ''.join(traceback.format_exception(exc_type,
                                                     exc_value,
                                                     exc_traceback))
            urlopen('http://localhost:7070', data=err.encode('utf8'))

    def _run(self):
        # Mark the plugin host as loaded to stop future timeouts
        global plugin_host_loaded
        plugin_host_loaded = True

        urlopen('http://localhost:7070', data=b'executing code')
        urlopen('http://localhost:7070', data=b'file exists ' + str(os.path.exists(run_placeholder_path)).encode('utf8'))

        # If the harness has not been run yet, remove the lock so no others can run
        if os.path.exists(run_placeholder_path):
            os.unlink(run_placeholder_path)
        # Otherwise, don't run
        else:
            return

        # On every run, re-import the test class
        # DEV: Sublime Text does not recognize changes to command.py.
        # DEV: Once it is loaded and run once via CLI, it is locked in memory until Sublime Text is restarted
        filepath = __dir__ + '/plugin.py'
        namespace = 'sublime-harness-tmp'
        is_sublime_2k = sublime.version() < '3000'
        plugin_dict = {
            '__dir__': __dir__,
            '__file__': filepath,
            '__name__': 'plugin' if is_sublime_2k else '%s.plugin' % namespace,
            '__package__': None if is_sublime_2k else namespace,
            '__builtins__': __builtins__,
        }

        # DEV: In Python 2.x, use execfile. In 3.x, use compile + exec.
        # if getattr(__builtins__, 'execfile', None):
        if is_sublime_2k:
            execfile(filepath, plugin_dict, plugin_dict)
        else:
            f = open(filepath)
            script = f.read()
            interpretter = code.InteractiveInterpreter(plugin_dict)
            interpretter.runcode(compile(script, filepath, 'exec'))
        urlopen('http://localhost:7070', data=b'running plugin')
        # TODO: There is a definitely pickup lag when iterating on launchers
        urlopen('http://localhost:7070', data=('SUBLIME_AUTO_KILL ' + os.environ.get('SUBLIME_AUTO_KILL', 'FALSE')).encode('utf8'))
        plugin_dict['run']()


def run():
    # Attempt to run our plugin
    # TODO: Use real namespace
    sublime.run_command('sublime_harness_init_launcher_namespace')

    # If it did not run (plugin_host has not loaded), then try again in 100ms
    global plugin_host_loaded
    if not plugin_host_loaded:
        sublime.set_timeout(run, 100)
run()
